node/src/components/consensus/highway_core/state/tests.rs

```diff
@@ -188,30 +194,85 @@ fn add_unit() -> Result<(), AddUnitError<TestContext>> {
+    let unit = SignedWireUnit::new(wunit.clone().into_hashed(), &BOB_SEC);
+    let maybe_err = state.add_unit(unit, panorama).err().map(unit_err);
     assert_eq!(Some(UnitError::ThreeUnitsInRound), maybe_err);
+    wunit.timestamp = 65.into();
+
+    // Your sequence number must match the panorama.
+    wunit.seq_number = 3;
+    let unit = SignedWireUnit::new(wunit.clone().into_hashed(), &BOB_SEC);
+    let maybe_err = state.pre_validate_unit(&unit, None).err();
+    assert_eq!(Some(UnitError::SequenceNumber), maybe_err);
+    wunit.seq_number = 2;
+
+    // If a hash-based panorama is included it must match the sequence number-based one.
+    let unit = SignedWireUnit::new(wunit.into_hashed(), &BOB_SEC);
+    let maybe_err = state
+        .add_unit(unit.clone(), panorama!(a0, b1, c0))
+        .err()
+        .map(unit_err);
+    assert_eq!(Some(UnitError::PanoramaMismatch(ALICE)), maybe_err);
+    let maybe_err = state
+        .add_unit(unit.clone(), panorama!(N, b1, c0, N))
+        .err()
+        .map(unit_err);
+    assert_eq!(Some(UnitError::PanoramaLength(4)), maybe_err);
+    let maybe_err = state
+        .add_unit(unit.clone(), panorama!(a0, b1, c0))
+        .err()
+        .map(unit_err);
+    assert_eq!(Some(UnitError::PanoramaMismatch(ALICE)), maybe_err);
+
+    // Your previous unit must match the panorama.
+    let maybe_err = state
+        .add_unit(unit, panorama!(a0, N, c0))
+        .err()
+        .map(unit_err);
+    assert_eq!(Some(UnitError::PreviousUnit), maybe_err);
 
     // Inconsistent panorama: If you see b1, you have to see c0, too.
     let maybe_err = add_unit!(state, CAROL, None; N, b1, N).err().map(unit_err);
     assert_eq!(Some(UnitError::InconsistentPanorama(BOB)), maybe_err);
-    // And you can't make the round exponent too small
+
+    // You can't make the round exponent too small.
     let maybe_err = add_unit!(state, CAROL, 50, 5u8, None; N, b1, c0)
         .err()
         .map(unit_err);
     assert_eq!(Some(UnitError::RoundLengthExpChangedWithinRound), maybe_err);
-    // And you can't make the round exponent too big
+
+    // And you can't make the round exponent too big.
     let maybe_err = add_unit!(state, CAROL, 50, 40u8, None; N, b1, c0)
         .err()
         .map(unit_err);
     assert_eq!(Some(UnitError::RoundLengthExpGreaterThanMaximum), maybe_err);
+
     // After the round from 48 to 64 has ended, the exponent can change.
     let c1 = add_unit!(state, CAROL, 65, 5u8, None; N, b1, c0)?;
 
+    // If you don't cite any other units, you must propose a block.
+    let maybe_err = add_unit!(state, CAROL, 0, 4u8, None; N, N, N)
+        .err()
+        .map(unit_err);
+    assert_eq!(Some(UnitError::MissingBlock), maybe_err);
+
+    // The panorama length must be the number of validators.
+    let maybe_err = add_unit!(state, CAROL, 0, 4u8, 0xC; N, N, N, N)
+        .err()
+        .map(unit_err);
+    assert_eq!(Some(UnitError::PanoramaLength(4)), maybe_err);
```

Comment:
Isn't that the same as in https://github.com/casper-network/casper-node/pull/1550/files#diff-5edd9079fa700dab6deb47fd36ebe675615f67a4d00a3112e0cbaeac2d1a63c4R221-R225 ?